### ISSUE TITLE
Default values of functions not being found by coverage

### ISSUE THREAD

**Issue by BackEndTea** [infection/infection#101]
| Question    | Answer
| ------------| ---------------
| Infection version | 0.7.0
| Test Framework version | PHPUnit/ 6.5.-
| PHP version | 7.1.12
| Platform    | MacOS
| Github Repo | https://github.com/BackEndTea/Array-Meta

<details>
 <summary>Output with issue</summary>
 Given the following code

```php
    public function search($value, bool $strict = false)
    {
        if (($return =  \array_search($value, $this->items, $strict)) === false) {
            throw ValueNotFoundException::valueNotFound($value);
        }
        return $return;
    }

```

And the following test

```php
    public function testSearchDefaultsToNonStrictSearch()
    {
        $array = ['0', '1', '2', '3'];
        $meta = new ArrayMeta($array);
        $key = $meta->search(3);
        $this->assertSame(3, $key);
    }
```

I expected the mutation of $strict to true to be covered. 

However it tells me the mutation was not covered, as the code coverage doesn't look at the function signature, but only to the body of the function. So mutating default values always says it was not covered

</details>

**Comment by borNfreee** [comment]
Thank you for opening this issue.

Actually, this is how PHPUnit's Coverage works:

![coverage-bug](https://user-images.githubusercontent.com/3725595/34920123-99f151e4-f97e-11e7-8bc1-db33236073c8.png)

Notice that coverage does not have information about function signature line (it is white, not green), even if the function was executed. To leverage this issue and to make e.g. Public/Protected Visibility mutators work, we are searching any executed line inside the called function, and if they are found, we treat the line with function signature _as covered_.

<Details>

https://github.com/infection/infection/blob/0f06fb474d30cc978419b66ba34ec22d70684934/src/TestFramework/Coverage/CodeCoverageData.php#L87-L106

</Details>

Currently, we have two types of mutators:

* Function body mutator (almost all mutators)
* Function signature mutator (Public/Protected visibility mutators)

It looks like function body mutator in your repository that changes `true` to `false` also changes values in the method signature. And from the first look, we can just not allow mutating code in the function signature for this type of mutators. But, your example shows that this is a pretty valid case and we should support these mutations.

Probably, the API of the [base Mutator abstract class](https://github.com/infection/infection/blob/master/src/Mutator/Mutator.php#L14) should be changed. Instead of having two methods (`isFunctionBodyMutator()` and `isFunctionSignatureMutator()`) we can have remove them and always for each mutator search covered line in the function signature as well (as described above in `hasExecutedMethodOnLine()`).

To sum it up:

* Mutation in the function signature should not be marked as "Not covered" _if this method was executed_ (reuse `hasExecutedMethodOnLine()`)
* We should get rid of `isFunctionBodyMutator()` and `isFunctionSignatureMutator()` methods

**Comment by borNfreee** [comment]
Fixed in #123
